core: Fixed variables not being in scope for destroy -target on modules

kyhavlov opened the pull request:
Addresses the bug in #8146. The call to `selectTargetedNodes()` was only returning the resource and not the variable that it depended on to interpolate its count, so the variable got removed from the graph and wasn't in scope for later interpolations, causing the `unknown variable accessed` error.


@phinze

phinze commented on terraform/context_apply_test.go:
```diff
@@ -3119,6 +3119,91 @@ module.child:
+	}
+
+	h := new(HookRecordApplyOrder)
+	h.Active = true
```
Looks like this Hook is not used in this test.

phinze commented on terraform/context_apply_test.go:
```diff
@@ -3119,6 +3119,91 @@ module.child:
+			Targets: []string{"module.child"},
+		})
+
+		// First plan and apply a create operation
```
This comment isn't applicable here anymore, since this is the destroy.

phinze commented on terraform/context_apply_test.go:
```diff
@@ -3119,6 +3119,91 @@ module.child:
+			Providers: map[string]ResourceProviderFactory{
+				"aws": testProviderFuncFixed(p),
+			},
+		})
```
The above `WritePlan` / `ReadPlan` dance is meant to emulate writing out to a `tfplan` file. It's unnecessary here, since the bug reproduces w/o it.

phinze commented on terraform/transform_targets.go:
```diff
@@ -86,6 +86,14 @@ func (t *TargetsTransformer) selectTargetedNodes(
 			if t.Destroy {
 				deps, err = g.Descendents(v)
+
+				// Select any variables that we depend on in case we need them later for interpolating in count
```
Style nit: wrap comment at 80 columns.

kyhavlov commented:
@phinze addressed the comments and cleaned up the test. I was going add a test for the nested module case, but I hit the issue with Terraform not including child modules of a module when using -target, so that will have to be another PR.

phinze commented:
LGTM

matthughes commented:
Got the error this was supposed to fix on 0.7.4.  Upgraded to 0.7.6.  Now I get:

```
Error configuring: 1 error(s) occurred:

* Error reading data.template_file.master_config count: strconv.ParseInt: parsing "${var.master_count}": invalid syntax
```
